### ISSUE TITLE
XMonad.Actions.SpawnOn.getPPIDOf assumes wrong format of proc/*/stat files

### ISSUE THREAD

**Issue by johannesgerer** [xmonad/xmonad-contrib#726]
### Problem Description

manageHook crashes with `Prelude.read: no parse`, when trying to evaluate `XMonad.Actions.SpawnOn.getPPIDOf 1162897`:

```
cat /proc/1162897/stat
1162897 (tmux: server) S 1 1162897 1162897 0 -1 4194368 4578 25034 0 0 649 111 22 14 20 0 1 0 6731470 242098176 2747 18446744073709551615 4214784 4799021 140724568486368 0 0 0 0 3674116 134433283 0 0 0 17 3 0 0 0 0 0 5019376 5098152 12685312 140724568492670 140724568492736 140724568492736 140724568502201 0
```

### Steps to Reproduce

This is difficult to pull of, but the example makes it clear that following [code](https://hackage.haskell.org/package/xmonad-contrib-0.17.0/docs/src/XMonad.Actions.SpawnOn.html#getPPIDOf):

```
getPPIDOf :: ProcessID -> Maybe ProcessID
getPPIDOf thisPid =
    case unsafePerformIO . tryJust (guard . isDoesNotExistError) . readFile . printf "/proc/%d/stat" $ toInteger (traceShowId thisPid) of
      Left _         -> Nothing
      Right contents -> case lines contents of
                          []        -> Nothing
                          first : _ -> case words first of
     (BUG IN THIS LINE)                  _ : _ : _ : ppid : _ -> Just $ fromIntegral (read ppid :: Int)
                                         _                    -> Nothing

```
assumes that the 4th word is the ppid which is not correct, as seen example file above.

**Comment by geekosaur** [comment]
More precisely, the second field can have spaces, as shown in this case. Since all the other fields are numeric and it's also possible to put parentheses in a process name (and they don't have to match),m the correct way to parse the line is to find the final closing parenthesis and consider that the end of the second field.

**Comment by johannesgerer** [comment]
Yes, what seems to work on my system is:

```
                          first : _ -> case words (dropWhile (/= ')') first) of
                                         _ : _ : ppid : _ -> Just $ fromIntegral (read ppid :: Int)
```

**Comment by slotThe** [comment]
> Yes, what seems to work on my system is:
>
> ```
>                           first : _ -> case words (dropWhile (/= ')') first) of
>                                          _ : _ : ppid : _ -> Just $ fromIntegral (read ppid :: Int)
> ```

If this

> it's also possible to put parentheses in a process name (and they don't have to match)

is correct then we would probably want something like `takeWhileEnd (/= ')')` here, but other than that this look fine to me.  Would you like to open a PR?

**Comment by johannesgerer** [comment]
You are right. And I think the name field can also contain newlines, so `lines` is not a good idea either

**Comment by liskin** [comment]
I need this function for a fix of WindowSwallowing that I've ~been working on~ had open in my text editor for the past month so I'd suggest that in addition to fixing this (using `takeWhileEnd` on the whole contents of the file, as there can be newlines in the process name indeed) we move the function to a common place like `XMonad.Util.Run` and reuse it from SpawnOn and WindowSwallowing. There's a slight chance I'll get to it by the end of the weekend, but please do feel free to make a pull request with the fix yourself if you want to try contributing to xmonad.

**Comment by slotThe** [comment]
@johannesgerer friendly ping :)

**Comment by johannesgerer** [comment]
Sorry, I won't be able to make a PR for the foreseeable future.

**Comment by liskin** [comment]
https://github.com/xmonad/xmonad-contrib/pull/732

**Comment by liskin** [comment]
(`takeWhileEnd` wasn't in the bytestring shipped with GHC 8.8 and we claim compatibility with 8.4+ so I used `spanEnd` instead)
